server/server.go

```diff
@@ -0,0 +1,229 @@
+package server
+
+import (
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/gob"
+	"fmt"
+	"io/ioutil"
+	"math"
+	"os"
+	"path"
+	"search/index"
+	"search/logger"
+	"search/searcher"
+	"search/util"
+	"strconv"
+	"time"
+)
+
+// Server contains all the necessary information for a running server.
+type Server struct {
+	mountPoint string        // Mount point of the server
+	lenMS      int           // Length of the master secret in bytes
+	keyHalves  [][]byte      // The server-side keyhalves
+	salts      [][]byte      // The salts for deriving the keys for the PRFs
+	numFiles   int           // The number of files currently stored in the server.  This is used to determine the next docID.
+	size       uint64        // The number of slots in the bloom filter index
+	latency    time.Duration // The latency between the server and the client
+	bandwidth  int           // The bandwidth of the link betweem the server and the client (in bps)
+}
+
+// CreateServer initializes a server with `numClients` clients with a master
+// secret of length `lenMS`, and generate salts with length `lenSalt`.  The
+// number of salts is given by `r = -log2(fpRate)`, where `fpRate` is the
+// desired false positive rate of the system.  `mountPoint` determines where the
+// server files will be stored.
+func CreateServer(numClients, lenMS, lenSalt int, mountPoint string, fpRate float64, numUniqWords uint64) *Server {
+	s := new(Server)
+	masterSecret := make([]byte, lenMS)
+	rand.Read(masterSecret)
+	s.keyHalves = make([][]byte, numClients)
+	s.lenMS = lenMS
+	for i := 0; i < numClients; i++ {
+		h := sha256.New()
+		h.Write([]byte(strconv.Itoa(i)))
+		cksum := h.Sum(nil)
+		s.keyHalves[i] = util.XorBytes(masterSecret, cksum, lenMS)
+	}
+	r := int(math.Ceil(-math.Log2(fpRate)))
+	s.size = uint64(math.Ceil(float64(numUniqWords) * float64(r) / math.Log(2)))
+	s.salts = util.GenerateSalts(r, lenSalt)
+	s.numFiles = 0
+	s.mountPoint = mountPoint
+	s.writeToFile()
+	return s
+}
+
+// CreateServerWithLog behaves the same as `CreateServer`, except for that it
+// also sets the logging parameters for the server.
+func CreateServerWithLog(numClients, lenMS, lenSalt int, mountPoint string, fpRate float64, numUniqWords uint64, latency time.Duration, bandwidth int) *Server {
```

Comment:
It's a little weird to call these "logging parameters" (and to suffix the function with "WithLog"), since there are really network simulation parameters, which are conceptually different from logging.  Maybe "WithNetConfig" instead?